services/meta/data.go

```diff
@@ -1037,10 +1047,16 @@ func (sgi *ShardGroupInfo) marshal() *internal.ShardGroupInfo {
 }
 
 // unmarshal deserializes from a protobuf representation.
-func (sgi *ShardGroupInfo) unmarshal(pb *internal.ShardGroupInfo) {
+func (sgi *ShardGroupInfo) unmarshal(pb *internal.ShardGroupInfo) error {
 	sgi.ID = pb.GetID()
 	sgi.StartTime = UnmarshalTime(pb.GetStartTime())
-	sgi.EndTime = UnmarshalTime(pb.GetEndTime())
+	if pb.Duration != nil {
+		sgi.EndTime = sgi.StartTime.Add(time.Duration(pb.GetDuration()))
+	} else if pb.EndTime != nil {
+		sgi.EndTime = UnmarshalTime(pb.GetEndTime())
+	} else {
+		return errors.New("unable to unmarshal EndTime, missing both EndTime and Duration")
```

Comment:
Should this be a panic instead? It should never happen, but I feel like I have to account for this situation since both of the items are optional.